### DISCUSSION TITLE
Performance Regression

### DISCUSSION THREAD

**Discussion by alfonsocj** [reactwg/react-native-new-architecture#99] [Q&A] [answered]
Our team has found that the performance on the new architecture is not on par with the old one when a text heavy screen is mounted on the navigation stack.

I have a [reproducible example](https://github.com/alfonsocj/new-arch-perf-example) where I ran the tests. [Here's](https://docs.google.com/spreadsheets/d/1J79qVMst6aXOoMXvSJPOEyGOdjB5AKyAxybkSP7-mw0/edit?pli=1#gid=1408667178) a spreadsheet with the results. The raw data for the results can be found [here](https://github.com/alfonsocj/new-arch-perf-example/tree/main/perf-tests).

The following results are from a release version of the app using an iPhone 12 Pro Max.

## Results (times in ms)
|Event              |Old Arch Non Virtualized| Old Arch Virtualized| New Arch Non Virtualized    | New Arch Virtualized|
|-------------------|---------------|-----------|-------------------|-----------|
|js_navigation_1    |42.195         |35.59      |168.415            |28.088     |
|js_navigation_2    |27.866         |27.735     |176.643            |25.733     |
|native_navigation_1|29.586         |23.3       |126.065            |17.152     |
|native_navigation_2|31.795         |24.164     |178.757            |20.418     |
|500-views          |126.551        |135.619    |307.222            |171.328    |
|1000-views         |237.216        |235.402    |452.423            |313.391    |
|1500-views         |342.467        |352.763    |613.525            |463.766    |

### Non Virtualized (times in ms)
| event               | Old Arch | New Arch | Percentage Increase |
|---------------------|----------|----------|---------------------|
| js_navigation_1     | 42.195   | 168.415  | -299.13%            |
| js_navigation_2     | 27.866   | 176.643  | -533.90%            |
| native_navigation_1 | 29.586   | 126.065  | -326.10%            |
| native_navigation_2 | 31.795   | 178.757  | -462.22%            |
| 500-views           | 126.551  | 307.222  | -142.77%            |
| 1000-views          | 237.216  | 452.423  | -90.72%             |
| 1500-views          | 342.467  | 613.525  | -79.15%             |

### Virtualized (times in ms)
| event               | Old Arch | New Arch | Percentage Increase |
|---------------------|----------|----------|---------------------|
| js_navigation_1     | 35.59    | 28.088   | 21.08%              |
| js_navigation_2     | 27.735   | 25.733   | 7.22%               |
| native_navigation_1 | 23.3     | 17.152   | 26.39%              |
| native_navigation_2 | 24.164   | 20.418   | 15.50%              |
| 500-views           | 135.619  | 171.328  | -26.33%             |
| 1000-views          | 235.402  | 313.391  | -33.13%             |
| 1500-views          | 352.763  | 463.766  | -31.47%             |

Here's a video of the test performed:
https://user-images.githubusercontent.com/3791120/204658443-30e486cd-f7f1-4ffe-be58-c81ab2159025.MP4

Each test was performed 10 times and the results in the spreadsheet are the average of the events.

## Reproduction
https://github.com/alfonsocj/new-arch-perf-example

The results seem to be similar to previous discussions [#88](https://github.com/reactwg/react-native-new-architecture/discussions/88#discussioncomment-4044743), [#85](https://github.com/reactwg/react-native-new-architecture/discussions/85) where heavy text screens seem to perform better on the old architecture, even though those tests were performed on Android builds.

I created an issue on React Navigation since it could be related https://github.com/react-navigation/react-navigation/issues/11054.

**Comment by cortinico** [comment]
Thanks for sharing this @alfonsocj. Have you engaged with folks on React Navigation by any chance?

**Reply by alfonsocj** [reply]
Hello, nobody from React Navigation has replied to the issue I created. I'll try to ask them on the discord channel. Hopefully someone can reproduce with the example I provided.

**Comment by alfonsocj** [comment]
An update, I tested on [my playground app](https://github.com/alfonsocj/new-arch-perf-example) without React Navigation and I still see a considerable performance regression when there are a lot of nodes, so we can discard the idea of React Navigation causing the performance issue.

**Reply by cortinico** [reply]
Thanks for flagging. I'd like to pass this to one of my colleague which has more context than me, but is sadly currently on vacation. We'll have to wait a bit till he gets back 👍

**Reply by alfonsocj** [reply]
Perfect. Let me know if there's something else I can help with.

**Reply by alfonsocj** [reply]
Happy New Year! @cortinico, could you please pass this to your colleague if he's back from vacation? Thanks!

**Comment by alfonsocj** [comment]
Hello! @cortinico any progress/updates on this issue?

**Reply by cortinico** [reply]
cc @sammy-SC if you can provide any insights here

**Reply by alfonsocj** [reply]
@sammy-SC let me know if there's something I can help you with.

**Reply by sammy-SC** [reply]
Hello @alfonsocj 

thank you for putting this together. 

At what point are markers started and ended? I noticed in one of your examples it says "endPerformanceMarker" and is called from [onLayout event](https://github.com/alfonsocj/new-arch-perf-example/blob/main/screens/heavy-screen.tsx#LL74C53-L74C53). Event `onLayout` does not reflect when a view is painted on the screen. It could be that the old and new infrastructure dispatches the events at different points in the pipeline.
One way to know that a view was painted on the screen is to have a custom component that overrides `didMoveToWindow` method on `UIView` and check that `window` property is not nil.

**Comment by sammy-SC** [comment, marked as answer]
Hello @alfonsocj,

we spent some time analysing performance and summarised our findings in https://github.com/reactwg/react-native-new-architecture/discussions/123.

Could you please try to run your benchmarks on the [0.72.0-RC.1](https://github.com/facebook/react-native/releases/tag/v0.72.0-rc.1) and see what the performance is like? We've made some improvements.
